packages/cli/src/runners/runner-ws-server.ts

```diff
@@ -1,30 +1,78 @@
+import { TaskRunnersConfig } from '@n8n/config';
 import type { BrokerMessage, RunnerMessage } from '@n8n/task-runner';
+import { ApplicationError } from 'n8n-workflow';
 import { Service } from 'typedi';
 import type WebSocket from 'ws';
 
+import { Time } from '@/constants';
 import { Logger } from '@/logging/logger.service';
 
 import { DefaultTaskRunnerDisconnectAnalyzer } from './default-task-runner-disconnect-analyzer';
+import { RunnerLifecycleEvents } from './runner-lifecycle-events';
 import type {
 	DisconnectAnalyzer,
+	DisconnectReason,
 	TaskRunnerServerInitRequest,
 	TaskRunnerServerInitResponse,
 } from './runner-types';
 import { TaskBroker, type MessageCallback, type TaskRunner } from './task-broker.service';
 
 function heartbeat(this: WebSocket) {
+	console.log('Received heartbeat from task runner'); // @TODO: Remove later
 	this.isAlive = true;
 }
 
 @Service()
 export class TaskRunnerWsServer {
 	runnerConnections: Map<TaskRunner['id'], WebSocket> = new Map();
 
+	private heartbeatTimer: NodeJS.Timer | undefined;
+
 	constructor(
 		private readonly logger: Logger,
 		private readonly taskBroker: TaskBroker,
 		private disconnectAnalyzer: DefaultTaskRunnerDisconnectAnalyzer,
-	) {}
+		private readonly taskTunnersConfig: TaskRunnersConfig,
+		private readonly runnerLifecycleEvents: RunnerLifecycleEvents,
+	) {
+		this.startHeartbeatChecks();
+	}
+
+	private startHeartbeatChecks() {
+		const { heartbeatInterval } = this.taskTunnersConfig;
+
+		if (heartbeatInterval <= 0) {
+			throw new ApplicationError('Heartbeat interval must be greater than 0');
+		}
+
+		this.heartbeatTimer = setInterval(() => {
+			this.runnerConnections.forEach((connection) => {
+				if (!connection.isAlive) {
+					const runnerId = Array.from(this.runnerConnections.entries()).find(
+						([_, c]) => c === connection,
+					)?.[0];
+					if (runnerId) void this.removeConnection(runnerId, 'failed-heartbeat-check');
+
+					this.runnerLifecycleEvents.emit('runner:failed-heartbeat-check');
+					return;
+				}
+
+				connection.isAlive = false;
+				connection.ping();
+			});
```

Comment:
We have both `runnerId` and `WebSocket` available in the `Map`, so we don't need to search for it:

```suggestion
			for (const [runnerId, connection] of this.runnerConnections.entries()) {
				if (!connection.isAlive) {
					void this.removeConnection(runnerId, 'failed-heartbeat-check');
					this.runnerLifecycleEvents.emit('runner:failed-heartbeat-check');
					return;
				}

				connection.isAlive = false;
				connection.ping();
			}
```